src/librustc_mir/hair/mod.rs

```diff
@@ -213,10 +213,7 @@ pub enum ExprKind<'tcx> {
     },
     Repeat {
         value: ExprRef<'tcx>,
-        // FIXME(#29789): Add a separate hair::Constant<'tcx> so this could be more explicit about
-        // its contained data. Currently this should only contain expression of ExprKind::Literal
-        // kind.
-        count: ExprRef<'tcx>,
+        count: Constant<'tcx>,
```

Comment:
MIR visitor does not really walk through `hair::ExprKind`s, does it? There’s no pressing reason to visit the `ConstVal` repetition count either, were you put it into `repr::Rvalue::Repeat` as well – it can be inspected as part of the whole `Rvalue::Repeat` inside the `visit_rvalue` (similarly as it is done with the `visit_literal(Literal::Value)`).

The issues with using `ConstVal` I could think of are loss of span and constant type information (latter being not really important now, because repetition count is always `usize`, but might be important in the future, for e.g. analysis passes). With @oli-obk’s work `ConstVal` is expected to eventually become typed, so loss of type would also, eventually, be fixed in time, leaving us with the span problem.